On the Client side
---

Use the 'right' venv for the ipynb kernel (Might need to first `Python: Select Interpreter`, before the venv is proposed as an available kernel).

**Install the packages for the demo in the venv:**

If you already installed an environment, and just wanna check if you have all the required packages:

In [1]:
!pip freeze -r ../requirements.txt

clearml==1.16.4
clearml-agent==1.9.1
scikit-learn==1.5.2
pandas==2.2.2
# from task
joblib==1.4.2
matplotlib==3.9.2
numpy==2.1.1

# for ipynb
ipywidgets==8.1.5
## The following requirements were added by pip freeze:
asttokens==2.4.1
attrs==23.2.0
certifi==2024.8.30
charset-normalizer==3.3.2
comm==0.2.2
contourpy==1.3.0
cycler==0.12.1
debugpy==1.8.5
decorator==5.1.1
distlib==0.3.8
exceptiongroup==1.2.2
executing==2.1.0
fastjsonschema==2.20.0
filelock==3.16.0
fonttools==4.53.1
furl==2.1.3
idna==3.8
ipykernel==6.29.5
ipython==8.27.0
jedi==0.19.1
jsonschema==4.23.0
jsonschema-specifications==2023.12.1
jupyter_client==8.6.2
jupyter_core==5.7.2
jupyterlab_widgets==3.0.13
kiwisolver==1.4.7
matplotlib-inline==0.1.7
nbformat==5.10.4
nest-asyncio==1.6.0
orderedmultidict==1.0.1
packaging==24.1
parso==0.8.4
pathlib2==2.3.7.post1
pexpect==4.9.0
pillow==10.4.0
platformdirs==4.3.2
prompt_toolkit==3.0.47
psutil==5.9.8
ptyprocess==0.7.0
pure_eval==0.2.3
Pygments==2.18.0
PyJWT==2.8.0
pyparsing==3.1.4
pyt

If you notice any:
```bash
WARNING: Requirement file [../requirements.txt] contains pkg_name, but package 'pkg_name' is not installed
```
Then, you are missing packages that will be used in the demo at some points.

**To install the rquired packages:**

Either call:
```bash
make install-env
```
Or directly run:

In [ ]:
# %pip install -r requirements.txt

> <svg style="line-height: 1; height: 1.5em; width: 1.5em; vertical-align: middle;" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><!--!Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free Copyright 2024 Fonticons, Inc.--><path fill='grey' d="M256 32c14.2 0 27.3 7.5 34.5 19.8l216 368c7.3 12.4 7.3 27.7 .2 40.1S486.3 480 472 480L40 480c-14.3 0-27.6-7.7-34.7-20.1s-7-27.8 .2-40.1l216-368C228.7 39.5 241.8 32 256 32zm0 128c-13.3 0-24 10.7-24 24l0 112c0 13.3 10.7 24 24 24s24-10.7 24-24l0-112c0-13.3-10.7-24-24-24zm32 224a32 32 0 1 0 -64 0 32 32 0 1 0 64 0z"/></svg> WARNING:\
> Using:
> ```python
> !the_cmd
> ```
> Resulted in weird behaviours: agent present to closing down and could not take on any task from the queue
> Instead using:
> ```python
> %%bash
> the_cmd
> ```
> Resulted in the correct behaviours

# Exploring agents

TODO List:
*  Check out the storage locations

First, let's check if any agent is running:

In [2]:
%%bash
clearml-agent list

workers:
- company:
    id: d1bd92a3b039400cbafc60a7a5b1e52b
    name: clearml
  id: clearml-agent-bengio:gpu0
  ip: 172.25.1.1
  key: worker_d1bd92a3b039400cbafc60a7a5b1e52b_0f358a174884f2c34ef7def624d3e1a0_clearml-agent-bengio:gpu0
  last_activity_time: '2024-09-29T15:35:37.871218+00:00'
  last_report_time: '2024-09-29T15:35:37.871218+00:00'
  queues:
  - id: 6b55ab8e3be4482eb53acfed3e4037a5
    name: gpu_dithering
    num_tasks: 0
  register_time: '2024-09-29T10:25:05.394822+00:00'
  register_timeout: 600
  system_tags: []
  tags: []
  user:
    id: 0f358a174884f2c34ef7def624d3e1a0
    name: RR5555



With no agent running, you should get the following output:
```bash
workers: []
```


**To launch an agent:**\
*(with name `clearml-agent-bengio`, id `clearml-agent-bengio:gpu0`, on gpu 0 with queue `gpu_dithering`)*

Either call:
```bash
make launch-agent
```
Or directly call:

In [ ]:
%%bash
CLEARML_WORKER_NAME=clearml-agent-bengio CLEARML_WORKER_ID=clearml-agent-bengio:gpu0 clearml-agent daemon --detached --queue gpu_dithering --create-queue --gpus 0

<details>
<summary>Output</summary>

```bash
Current configuration (clearml_agent v1.9.1, location: /home/blissfox-server/clearml.conf):
----------------------
agent.worker_id = clearml-agent-bengio:gpu0
agent.worker_name = clearml-agent-bengio
agent.force_git_ssh_protocol = false
agent.python_binary = /usr/bin/python3.10
agent.package_manager.type = pip
agent.package_manager.pip_version.0 = <20.2 ; python_version < '3.10'
agent.package_manager.pip_version.1 = <22.3 ; python_version >\= '3.10' and python_version <\= '3.11'
agent.package_manager.pip_version.2 = >\=23,<24.3 ; python_version >\= '3.12'
agent.package_manager.system_site_packages = false
agent.package_manager.force_upgrade = false
agent.package_manager.conda_channels.0 = pytorch
agent.package_manager.conda_channels.1 = conda-forge
agent.package_manager.conda_channels.2 = nvidia
agent.package_manager.conda_channels.3 = defaults
agent.package_manager.priority_optional_packages.0 = pygobject
agent.package_manager.torch_nightly = false
agent.package_manager.poetry_files_from_repo_working_dir = false
agent.venvs_dir = /home/blissfox-server/.clearml/venvs-builds
agent.venvs_cache.max_entries = 10
agent.venvs_cache.free_space_threshold_gb = 2.0
agent.venvs_cache.path = ~/.clearml/venvs-cache
agent.vcs_cache.enabled = true
agent.vcs_cache.path = /home/blissfox-server/.clearml/vcs-cache
agent.venv_update.enabled = false
agent.pip_download_cache.enabled = true
agent.pip_download_cache.path = /home/blissfox-server/.clearml/pip-download-cache
agent.translate_ssh = true
agent.reload_config = false
agent.docker_pip_cache = /home/blissfox-server/.clearml/pip-cache
agent.docker_apt_cache = /home/blissfox-server/.clearml/apt-cache
agent.docker_force_pull = false
agent.default_docker.image = nvidia/cuda:11.0.3-cudnn8-runtime-ubuntu20.04
agent.default_docker.match_rules.0.image = python:3.6-bullseye
agent.default_docker.match_rules.0.arguments = --ipc\=host
agent.default_docker.match_rules.0.match.script.binary = python3.6$
agent.default_docker.match_rules.1.image = python:3.7-bullseye
agent.default_docker.match_rules.1.arguments = --ipc\=host
agent.default_docker.match_rules.1.match.script.binary = python3.7$
agent.default_docker.match_rules.2.image = python:3.8-bullseye
agent.default_docker.match_rules.2.arguments = --ipc\=host
agent.default_docker.match_rules.2.match.script.binary = python3.8$
agent.default_docker.match_rules.3.image = python:3.9-bullseye
agent.default_docker.match_rules.3.arguments = --ipc\=host
agent.default_docker.match_rules.3.match.script.binary = python3.9$
agent.default_docker.match_rules.4.image = python:3.10-bullseye
agent.default_docker.match_rules.4.arguments = --ipc\=host
agent.default_docker.match_rules.4.match.script.binary = python3.10$
agent.default_docker.match_rules.5.image = python:3.11-bullseye
agent.default_docker.match_rules.5.arguments = --ipc\=host
agent.default_docker.match_rules.5.match.script.binary = python3.11$
agent.default_docker.match_rules.6.image = python:3.12-bullseye
agent.default_docker.match_rules.6.arguments = --ipc\=host
agent.default_docker.match_rules.6.match.script.binary = python3.12$
agent.enable_task_env = false
agent.sanitize_config_printout = ****
agent.hide_docker_command_env_vars.enabled = true
agent.hide_docker_command_env_vars.parse_embedded_urls = true
agent.abort_callback_max_timeout = 1800
agent.docker_internal_mounts.sdk_cache = /clearml_agent_cache
agent.docker_internal_mounts.apt_cache = /var/cache/apt/archives
agent.docker_internal_mounts.ssh_folder = ~/.ssh
agent.docker_internal_mounts.ssh_ro_folder = /.ssh
agent.docker_internal_mounts.pip_cache = /root/.cache/pip
agent.docker_internal_mounts.poetry_cache = /root/.cache/pypoetry
agent.docker_internal_mounts.vcs_cache = /root/.clearml/vcs-cache
agent.docker_internal_mounts.venvs_cache = /root/.clearml/venvs-cache
agent.docker_internal_mounts.venv_build = ~/.clearml/venvs-builds
agent.docker_internal_mounts.pip_download = /root/.clearml/pip-download-cache
agent.apply_environment = true
agent.apply_files = true
agent.custom_build_script = 
agent.disable_task_docker_override = false
agent.git_user = RR5555
agent.git_pass = ****
agent.default_python = 3.10
agent.cuda_version = 122
agent.cudnn_version = 0
sdk.storage.cache.default_base_dir = ~/.clearml/cache
sdk.storage.cache.size.min_free_bytes = 10GB
sdk.storage.direct_access.0.url = file://*
sdk.metrics.file_history_size = 100
sdk.metrics.matplotlib_untitled_history_size = 100
sdk.metrics.images.format = JPEG
sdk.metrics.images.quality = 87
sdk.metrics.images.subsampling = 0
sdk.metrics.tensorboard_single_series_per_graph = false
sdk.network.metrics.file_upload_threads = 4
sdk.network.metrics.file_upload_starvation_warning_sec = 120
sdk.network.iteration.max_retries_on_server_error = 5
sdk.network.iteration.retry_backoff_factor_sec = 10
sdk.network.file_upload_retries = 3
sdk.aws.s3.key = 
sdk.aws.s3.secret = ****
sdk.aws.s3.region = 
sdk.aws.s3.use_credentials_chain = false
sdk.aws.boto3.pool_connections = 512
sdk.aws.boto3.max_multipart_concurrency = 16
sdk.aws.boto3.multipart_threshold = 8388608
sdk.aws.boto3.multipart_chunksize = 8388608
sdk.log.null_log_propagate = false
sdk.log.task_log_buffer_capacity = 66
sdk.log.disable_urllib3_info = true
sdk.development.task_reuse_time_window_in_hours = 72.0
sdk.development.vcs_repo_detect_async = true
sdk.development.store_uncommitted_code_diff = true
sdk.development.support_stopping = true
sdk.development.default_output_uri = 
sdk.development.force_analyze_entire_repo = false
sdk.development.suppress_update_message = false
sdk.development.detect_with_pip_freeze = false
sdk.development.worker.report_period_sec = 2
sdk.development.worker.ping_period_sec = 30
sdk.development.worker.log_stdout = true
sdk.development.worker.report_global_mem_used = false
sdk.development.worker.report_event_flush_threshold = 100
sdk.development.worker.console_cr_flush_period = 10
sdk.apply_environment = false
sdk.apply_files = false
api.version = 1.5
api.verify_certificate = true
api.default_version = 1.5
api.http.max_req_size = 15728640
api.http.retries.total = 240
api.http.retries.connect = 240
api.http.retries.read = 240
api.http.retries.redirect = 240
api.http.retries.status = 240
api.http.retries.backoff_factor = 1.0
api.http.retries.backoff_max = 120.0
api.http.wait_on_maintenance_forever = true
api.http.pool_maxsize = 512
api.http.pool_connections = 512
api.auth.token_expiration_threshold_sec = ****
api.api_server = https://api.clearml.blissfox.xyz
api.web_server = https://app.clearml.blissfox.xyz
api.files_server = https://files.clearml.blissfox.xyz
api.credentials.access_key = EUAQEJNSXKQU1UN76PMC
api.credentials.secret_key = ****
api.host = https://api.clearml.blissfox.xyz

Worker "clearml-agent-bengio:gpu0" - Listening to queues:
+----------------------------------+---------------+-------+
| id                               | name          | tags  |
+----------------------------------+---------------+-------+
| 6b55ab8e3be4482eb53acfed3e4037a5 | gpu_dithering |       |
+----------------------------------+---------------+-------+

Running CLEARML-AGENT daemon in background mode, writing stdout/stderr to /tmp/.clearml_agent_daemon_outho7nopyx.txt
```

</details>

Note:
```bash
Running CLEARML-AGENT daemon in background mode, writing stdout/stderr to /tmp/.clearml_agent_daemon_outho7nopyx.txt
```

<details>
<summary>About the cmd & arguments</summary>

From [ClearML Agent CLI](https://clear.ml/docs/latest/docs/clearml_agent/clearml_agent_ref/):
|||
|:---|:---|
|`daemon`|command to spin up an agent on any machine|
|`--create-queue`|If the queue name provided with `--queue` does not exist in the server, create it on-the-fly and use it.|
|`--detached`|Run agent in the background. The `clearml-agent` command returns immediately.|
|`--gpus`|If running in Docker mode (see the `--docker` option), specify the active GPUs for the Docker containers to use. These are the same GPUs set in the `NVIDIA_VISIBLE_DEVICES` environment variable. For example:<ul><li>`--gpus 0`</li><li>`--gpu 0,1,2`</li><li>`--gpus all`</li></ul>|
|`--queue`|Specify the queues which the worker is listening to. The values can be any combination of:<ul><li>One or more queue IDs</li><li>One or more queue names</li><li>`default` indicating the default queue</li></ul>
</details>

<details>
<summary>On the GUI:</summary>

* Before:\
    ![](./../img/Workers_null.png)
* After:\
    ![](./../img/Workers_bengio0_0.png)\
    ![](./../img/Workers_bengio0_1.png)

</details>

Let's check out again the agent available from the CLI:

In [ ]:
%%bash
clearml-agent list

workers:
- company:
    id: d1bd92a3b039400cbafc60a7a5b1e52b
    name: clearml
  id: clearml-agent-bengio:gpu0
  ip: 172.25.1.1
  key: worker_d1bd92a3b039400cbafc60a7a5b1e52b_0f358a174884f2c34ef7def624d3e1a0_clearml-agent-bengio:gpu0
  last_activity_time: '2024-09-28T10:27:23.587122+00:00'
  last_report_time: '2024-09-28T10:27:23.587122+00:00'
  queues:
  - id: 6b55ab8e3be4482eb53acfed3e4037a5
    name: gpu_dithering
    num_tasks: 0
  register_time: '2024-09-28T10:27:20.249917+00:00'
  register_timeout: 600
  system_tags: []
  tags: []
  user:
    id: 0f358a174884f2c34ef7def624d3e1a0
    name: RR5555



<details>
<summary>Output</summary>

```bash
workers:
- company:
    id: d1bd92a3b039400cbafc60a7a5b1e52b
    name: clearml
  id: clearml-agent-bengio:gpu0
  ip: 172.25.1.1
  key: worker_d1bd92a3b039400cbafc60a7a5b1e52b_0f358a174884f2c34ef7def624d3e1a0_clearml-agent-bengio:gpu0
  last_activity_time: '2024-09-16T11:46:44.589928+00:00'
  queues:
  - id: 6b55ab8e3be4482eb53acfed3e4037a5
    name: gpu_dithering
    num_tasks: 0
  register_time: '2024-09-16T11:46:44.589928+00:00'
  register_timeout: 600
  system_tags: []
  tags: []
  user:
    id: 0f358a174884f2c34ef7def624d3e1a0
    name: RR5555
```
</details>

In [ ]:
%%bash
docker ps

In [8]:
%%bash
ls -a /tmp/ | grep -i 'clearml\|requirements'
# OR equivalently:
# ls -a /tmp/ | grep -E -i 'clearml|requirements'

.clearmlagent_167497_9u1qoi1a.tmp
.clearml_agent.3tikdf0t.cfg
.clearml_agent_daemon_outho7nopyx.txt
requirements_uu2a4f6i.txt


With one agent running, we get smth like:
```bash
.clearmlagent_167497_9u1qoi1a.tmp
.clearml_agent.3tikdf0t.cfg
.clearml_agent_daemon_outho7nopyx.txt
requirements_uu2a4f6i.txt
```
With no agent running, we get smth like:
```bash
requirements_uu2a4f6i.txt
```

In [ ]:
%%bash
cat /tmp/.clearml_agent_daemon_*.txt

<details>
<summary>Output</summary>

```bash
Worker "clearml-agent-bengio:gpu0" :
Current configuration (clearml_agent v1.9.1, location: /home/blissfox-server/clearml.conf):
----------------------
agent.worker_id = clearml-agent-bengio:gpu0
agent.worker_name = clearml-agent-bengio
agent.force_git_ssh_protocol = false
agent.python_binary = /usr/bin/python3.10
agent.package_manager.type = pip
agent.package_manager.pip_version.0 = <20.2 ; python_version < '3.10'
agent.package_manager.pip_version.1 = <22.3 ; python_version >\= '3.10' and python_version <\= '3.11'
agent.package_manager.pip_version.2 = >\=23,<24.3 ; python_version >\= '3.12'
agent.package_manager.system_site_packages = false
agent.package_manager.force_upgrade = false
agent.package_manager.conda_channels.0 = pytorch
agent.package_manager.conda_channels.1 = conda-forge
agent.package_manager.conda_channels.2 = nvidia
agent.package_manager.conda_channels.3 = defaults
agent.package_manager.priority_optional_packages.0 = pygobject
agent.package_manager.torch_nightly = false
agent.package_manager.poetry_files_from_repo_working_dir = false
agent.venvs_dir = /home/blissfox-server/.clearml/venvs-builds
agent.venvs_cache.max_entries = 10
agent.venvs_cache.free_space_threshold_gb = 2.0
agent.venvs_cache.path = ~/.clearml/venvs-cache
agent.vcs_cache.enabled = true
agent.vcs_cache.path = /home/blissfox-server/.clearml/vcs-cache
agent.venv_update.enabled = false
agent.pip_download_cache.enabled = true
agent.pip_download_cache.path = /home/blissfox-server/.clearml/pip-download-cache
agent.translate_ssh = true
agent.reload_config = false
agent.docker_pip_cache = /home/blissfox-server/.clearml/pip-cache
agent.docker_apt_cache = /home/blissfox-server/.clearml/apt-cache
agent.docker_force_pull = false
agent.default_docker.image = nvidia/cuda:11.0.3-cudnn8-runtime-ubuntu20.04
agent.default_docker.match_rules.0.image = python:3.6-bullseye
agent.default_docker.match_rules.0.arguments = --ipc\=host
agent.default_docker.match_rules.0.match.script.binary = python3.6$
agent.default_docker.match_rules.1.image = python:3.7-bullseye
agent.default_docker.match_rules.1.arguments = --ipc\=host
agent.default_docker.match_rules.1.match.script.binary = python3.7$
agent.default_docker.match_rules.2.image = python:3.8-bullseye
agent.default_docker.match_rules.2.arguments = --ipc\=host
agent.default_docker.match_rules.2.match.script.binary = python3.8$
agent.default_docker.match_rules.3.image = python:3.9-bullseye
agent.default_docker.match_rules.3.arguments = --ipc\=host
agent.default_docker.match_rules.3.match.script.binary = python3.9$
agent.default_docker.match_rules.4.image = python:3.10-bullseye
agent.default_docker.match_rules.4.arguments = --ipc\=host
agent.default_docker.match_rules.4.match.script.binary = python3.10$
agent.default_docker.match_rules.5.image = python:3.11-bullseye
agent.default_docker.match_rules.5.arguments = --ipc\=host
agent.default_docker.match_rules.5.match.script.binary = python3.11$
agent.default_docker.match_rules.6.image = python:3.12-bullseye
agent.default_docker.match_rules.6.arguments = --ipc\=host
agent.default_docker.match_rules.6.match.script.binary = python3.12$
agent.enable_task_env = false
agent.sanitize_config_printout = ****
agent.hide_docker_command_env_vars.enabled = true
agent.hide_docker_command_env_vars.parse_embedded_urls = true
agent.abort_callback_max_timeout = 1800
agent.docker_internal_mounts.sdk_cache = /clearml_agent_cache
agent.docker_internal_mounts.apt_cache = /var/cache/apt/archives
agent.docker_internal_mounts.ssh_folder = ~/.ssh
agent.docker_internal_mounts.ssh_ro_folder = /.ssh
agent.docker_internal_mounts.pip_cache = /root/.cache/pip
agent.docker_internal_mounts.poetry_cache = /root/.cache/pypoetry
agent.docker_internal_mounts.vcs_cache = /root/.clearml/vcs-cache
agent.docker_internal_mounts.venvs_cache = /root/.clearml/venvs-cache
agent.docker_internal_mounts.venv_build = ~/.clearml/venvs-builds
agent.docker_internal_mounts.pip_download = /root/.clearml/pip-download-cache
agent.apply_environment = true
agent.apply_files = true
agent.custom_build_script = 
agent.disable_task_docker_override = false
agent.git_user = RR5555
agent.git_pass = ****
agent.default_python = 3.10
agent.cuda_version = 122
agent.cudnn_version = 0
sdk.storage.cache.default_base_dir = ~/.clearml/cache
sdk.storage.cache.size.min_free_bytes = 10GB
sdk.storage.direct_access.0.url = file://*
sdk.metrics.file_history_size = 100
sdk.metrics.matplotlib_untitled_history_size = 100
sdk.metrics.images.format = JPEG
sdk.metrics.images.quality = 87
sdk.metrics.images.subsampling = 0
sdk.metrics.tensorboard_single_series_per_graph = false
sdk.network.metrics.file_upload_threads = 4
sdk.network.metrics.file_upload_starvation_warning_sec = 120
sdk.network.iteration.max_retries_on_server_error = 5
sdk.network.iteration.retry_backoff_factor_sec = 10
sdk.network.file_upload_retries = 3
sdk.aws.s3.key = 
sdk.aws.s3.secret = ****
sdk.aws.s3.region = 
sdk.aws.s3.use_credentials_chain = false
sdk.aws.boto3.pool_connections = 512
sdk.aws.boto3.max_multipart_concurrency = 16
sdk.aws.boto3.multipart_threshold = 8388608
sdk.aws.boto3.multipart_chunksize = 8388608
sdk.log.null_log_propagate = false
sdk.log.task_log_buffer_capacity = 66
sdk.log.disable_urllib3_info = true
sdk.development.task_reuse_time_window_in_hours = 72.0
sdk.development.vcs_repo_detect_async = true
sdk.development.store_uncommitted_code_diff = true
sdk.development.support_stopping = true
sdk.development.default_output_uri = 
sdk.development.force_analyze_entire_repo = false
sdk.development.suppress_update_message = false
sdk.development.detect_with_pip_freeze = false
sdk.development.worker.report_period_sec = 2
sdk.development.worker.ping_period_sec = 30
sdk.development.worker.log_stdout = true
sdk.development.worker.report_global_mem_used = false
sdk.development.worker.report_event_flush_threshold = 100
sdk.development.worker.console_cr_flush_period = 10
sdk.apply_environment = false
sdk.apply_files = false
api.version = 1.5
api.verify_certificate = true
api.default_version = 1.5
api.http.max_req_size = 15728640
api.http.retries.total = 240
api.http.retries.connect = 240
api.http.retries.read = 240
api.http.retries.redirect = 240
api.http.retries.status = 240
api.http.retries.backoff_factor = 1.0
api.http.retries.backoff_max = 120.0
api.http.wait_on_maintenance_forever = true
api.http.pool_maxsize = 512
api.http.pool_connections = 512
api.auth.token_expiration_threshold_sec = ****
api.api_server = https://api.clearml.blissfox.xyz
api.web_server = https://app.clearml.blissfox.xyz
api.files_server = https://files.clearml.blissfox.xyz
api.credentials.access_key = EUAQEJNSXKQU1UN76PMC
api.credentials.secret_key = ****
api.host = https://api.clearml.blissfox.xyz

+----------------------------------+---------------+-------+
| id                               | name          | tags  |
+----------------------------------+---------------+-------+
| 6b55ab8e3be4482eb53acfed3e4037a5 | gpu_dithering |       |
+----------------------------------+---------------+-------+

Starting infinite task polling loop...
```

</details>

In [ ]:
%%bash
cat /tmp/.clearml_agent.*.cfg

<details>
<summary>Output</summary>

```cfg
{
    "agent": {
        "worker_id": "clearml-agent-bengio:gpu0",
        "worker_name": "clearml-agent-bengio",
        "force_git_ssh_protocol": false,
        "python_binary": "/usr/bin/python3.10",
        "package_manager": {
            "type": "pip",
            "pip_version": [
                "<20.2 ; python_version < '3.10'",
                "<22.3 ; python_version >= '3.10' and python_version <= '3.11'",
                ">=23,<24.3 ; python_version >= '3.12'"
            ],
            "system_site_packages": false,
            "force_upgrade": false,
            "conda_channels": [
                "pytorch",
                "conda-forge",
                "nvidia",
                "defaults"
            ],
            "priority_optional_packages": [
                "pygobject"
            ],
            "torch_nightly": false,
            "poetry_files_from_repo_working_dir": false,
            "extra_index_url": []
        },
        "venvs_dir": "/home/blissfox-server/.clearml/venvs-builds",
        "venvs_cache": {
            "max_entries": 10,
            "free_space_threshold_gb": 2.0,
            "path": "~/.clearml/venvs-cache"
        },
        "vcs_cache": {
            "enabled": true,
            "path": "/home/blissfox-server/.clearml/vcs-cache"
        },
        "venv_update": {
            "enabled": false
        },
        "pip_download_cache": {
            "enabled": true,
            "path": "/home/blissfox-server/.clearml/pip-download-cache"
        },
        "translate_ssh": true,
        "reload_config": false,
        "docker_pip_cache": "/home/blissfox-server/.clearml/pip-cache",
        "docker_apt_cache": "/home/blissfox-server/.clearml/apt-cache",
        "docker_force_pull": false,
        "default_docker": {
            "image": "nvidia/cuda:11.0.3-cudnn8-runtime-ubuntu20.04",
            "match_rules": [
                {
                    "image": "python:3.6-bullseye",
                    "arguments": "--ipc=host",
                    "match": {
                        "script": {
                            "binary": "python3.6$"
                        }
                    }
                },
                {
                    "image": "python:3.7-bullseye",
                    "arguments": "--ipc=host",
                    "match": {
                        "script": {
                            "binary": "python3.7$"
                        }
                    }
                },
                {
                    "image": "python:3.8-bullseye",
                    "arguments": "--ipc=host",
                    "match": {
                        "script": {
                            "binary": "python3.8$"
                        }
                    }
                },
                {
                    "image": "python:3.9-bullseye",
                    "arguments": "--ipc=host",
                    "match": {
                        "script": {
                            "binary": "python3.9$"
                        }
                    }
                },
                {
                    "image": "python:3.10-bullseye",
                    "arguments": "--ipc=host",
                    "match": {
                        "script": {
                            "binary": "python3.10$"
                        }
                    }
                },
                {
                    "image": "python:3.11-bullseye",
                    "arguments": "--ipc=host",
                    "match": {
                        "script": {
                            "binary": "python3.11$"
                        }
                    }
                },
                {
                    "image": "python:3.12-bullseye",
                    "arguments": "--ipc=host",
                    "match": {
                        "script": {
                            "binary": "python3.12$"
                        }
                    }
                }
            ]
        },
        "enable_task_env": false,
        "sanitize_config_printout": {
            "hide_secrets": [
                "^sanitize_config_printout$",
                "secret",
                "pass",
                "token",
                "account_key",
                "contents"
            ],
            "hide_secrets_recursive": [
                "^environment$"
            ],
            "dont_hide_secrets": [
                "^enable_git_ask_pass$"
            ],
            "docker_commands": [
                "^extra_docker_arguments$"
            ],
            "urls": [
                "^extra_index_url$"
            ]
        },
        "hide_docker_command_env_vars": {
            "enabled": true,
            "extra_keys": [],
            "parse_embedded_urls": true
        },
        "abort_callback_max_timeout": 1800,
        "docker_internal_mounts": {
            "sdk_cache": "/clearml_agent_cache",
            "apt_cache": "/var/cache/apt/archives",
            "ssh_folder": "~/.ssh",
            "ssh_ro_folder": "/.ssh",
            "pip_cache": "/root/.cache/pip",
            "poetry_cache": "/root/.cache/pypoetry",
            "vcs_cache": "/root/.clearml/vcs-cache",
            "venvs_cache": "/root/.clearml/venvs-cache",
            "venv_build": "~/.clearml/venvs-builds",
            "pip_download": "/root/.clearml/pip-download-cache"
        },
        "apply_environment": true,
        "apply_files": true,
        "custom_build_script": "",
        "disable_task_docker_override": false,
        "git_user": "RR5555",
        "git_pass": "github_pat_**********************_***********************************************************",
        "default_python": "3.10",
        "cuda_version": "122",
        "cudnn_version": "0"
    },
    "sdk": {
        "storage": {
            "cache": {
                "default_base_dir": "~/.clearml/cache",
                "size": {
                    "min_free_bytes": "10GB"
                }
            },
            "direct_access": [
                {
                    "url": "file://*"
                }
            ]
        },
        "metrics": {
            "file_history_size": 100,
            "matplotlib_untitled_history_size": 100,
            "images": {
                "format": "JPEG",
                "quality": 87,
                "subsampling": 0
            },
            "tensorboard_single_series_per_graph": false
        },
        "network": {
            "metrics": {
                "file_upload_threads": 4,
                "file_upload_starvation_warning_sec": 120
            },
            "iteration": {
                "max_retries_on_server_error": 5,
                "retry_backoff_factor_sec": 10
            },
            "file_upload_retries": 3
        },
        "aws": {
            "s3": {
                "key": "",
                "secret": "",
                "region": "",
                "credentials": [],
                "use_credentials_chain": false,
                "extra_args": {}
            },
            "boto3": {
                "pool_connections": 512,
                "max_multipart_concurrency": 16,
                "multipart_threshold": 8388608,
                "multipart_chunksize": 8388608
            }
        },
        "google": {
            "storage": {}
        },
        "azure": {
            "storage": {}
        },
        "log": {
            "null_log_propagate": false,
            "task_log_buffer_capacity": 66,
            "disable_urllib3_info": true
        },
        "development": {
            "task_reuse_time_window_in_hours": 72.0,
            "vcs_repo_detect_async": true,
            "store_uncommitted_code_diff": true,
            "support_stopping": true,
            "default_output_uri": "",
            "force_analyze_entire_repo": false,
            "suppress_update_message": false,
            "detect_with_pip_freeze": false,
            "worker": {
                "report_period_sec": 2,
                "ping_period_sec": 30,
                "log_stdout": true,
                "report_global_mem_used": false,
                "report_event_flush_threshold": 100,
                "console_cr_flush_period": 10
            },
            "log_os_environments": []
        },
        "apply_environment": false,
        "apply_files": false
    },
    "logging": {
        "version": 1,
        "loggers": {
            "urllib3": {
                "level": "ERROR"
            }
        }
    },
    "api": {
        "version": 1.5,
        "verify_certificate": true,
        "default_version": 1.5,
        "http": {
            "max_req_size": 15728640,
            "retries": {
                "total": 240,
                "connect": 240,
                "read": 240,
                "redirect": 240,
                "status": 240,
                "backoff_factor": 1.0,
                "backoff_max": 120.0
            },
            "wait_on_maintenance_forever": true,
            "pool_maxsize": 512,
            "pool_connections": 512
        },
        "auth": {
            "token_expiration_threshold_sec": 43200
        },
        "api_server": "https://api.clearml.blissfox.xyz",
        "web_server": "https://app.clearml.blissfox.xyz",
        "files_server": "https://files.clearml.blissfox.xyz",
        "credentials": {
            "access_key": "EUAQEJNSXKQU1UN76PMC",
            "secret_key": "**************************************************"
        },
        "host": "https://api.clearml.blissfox.xyz"
    },
    "env": "default"
}
```

</details>

Some secrets actually appear in clear:
* `api`>`credentials`>`secret_key`
* `agent`>`git_pass`

In [ ]:
%%bash
cat /tmp/.clearmlagent_*.tmp

<details>
<summary>Output</summary>

```bash
clearml-agent-bengio:gpu0
0
```

</details>

To populate the log of the agent, we can launch the following exp.:

In [18]:
%%bash
clearml-task --project Demo --name clearml_test --script test.py --repo https://github.com/RR5555/clearml_demo.git --branch main --queue gpu_dithering --docker rr5555/dith:cuda12.0-agent

ClearML launch - launch any codebase on remote machine running clearml-agent
Creating new task
Repository Detected
{
  "repository": "https://github.com/RR5555/clearml_demo.git",
  "branch": "main",
  "working_dir": ".",
  "entry_point": "test.py"
}
Base docker image: {'image': 'rr5555/dith:cuda12.0-agent', 'args': None, 'bash_script': None}
New task created id=4f2118bf61c64795b89a3cd26b424fbe
Task id=4f2118bf61c64795b89a3cd26b424fbe sent for execution on queue gpu_dithering
Execution log at: https://app.clearml.blissfox.xyz/projects/ce6d125a264149bc89e8e62fb1dd0d7a/experiments/4f2118bf61c64795b89a3cd26b424fbe/output/log


<details>
<summary>Output</summary>

```bash
ClearML launch - launch any codebase on remote machine running clearml-agent
Creating new task
Repository Detected
{
  "repository": "https://github.com/RR5555/clearml_demo.git",
  "branch": "main",
  "working_dir": ".",
  "entry_point": "test.py"
}
Base docker image: {'image': 'rr5555/dith:cuda12.0-agent', 'args': None, 'bash_script': None}
New task created id=4f2118bf61c64795b89a3cd26b424fbe
Task id=4f2118bf61c64795b89a3cd26b424fbe sent for execution on queue gpu_dithering
Execution log at: https://app.clearml.blissfox.xyz/projects/ce6d125a264149bc89e8e62fb1dd0d7a/experiments/4f2118bf61c64795b89a3cd26b424fbe/output/log
```

</details>

In [ ]:
%%bash
cat /tmp/.clearml_agent_daemon_*.txt

<details>
<summary>Output</summary>

```bash
task 4f2118bf61c64795b89a3cd26b424fbe pulled from 6b55ab8e3be4482eb53acfed3e4037a5 by worker clearml-agent-bengio:gpu0
Running task '4f2118bf61c64795b89a3cd26b424fbe'
Storing stdout and stderr log to '/tmp/.clearml_agent_out.np498ozg.txt', '/tmp/.clearml_agent_out.np498ozg.txt'
Current configuration (clearml_agent v1.9.1, location: /tmp/.clearml_agent.3tikdf0t.cfg):
----------------------
agent.worker_id = clearml-agent-bengio:gpu0
agent.worker_name = clearml-agent-bengio
agent.force_git_ssh_protocol = false
agent.python_binary = /usr/bin/python3.10
agent.package_manager.type = pip
agent.package_manager.pip_version.0 = <20.2 ; python_version < '3.10'
agent.package_manager.pip_version.1 = <22.3 ; python_version >\= '3.10' and python_version <\= '3.11'
agent.package_manager.pip_version.2 = >\=23,<24.3 ; python_version >\= '3.12'
agent.package_manager.system_site_packages = false
agent.package_manager.force_upgrade = false
agent.package_manager.conda_channels.0 = pytorch
agent.package_manager.conda_channels.1 = conda-forge
agent.package_manager.conda_channels.2 = nvidia
agent.package_manager.conda_channels.3 = defaults
agent.package_manager.priority_optional_packages.0 = pygobject
agent.package_manager.torch_nightly = false
agent.package_manager.poetry_files_from_repo_working_dir = false
agent.venvs_dir = /home/blissfox-server/.clearml/venvs-builds
agent.venvs_cache.max_entries = 10
agent.venvs_cache.free_space_threshold_gb = 2.0
agent.venvs_cache.path = ~/.clearml/venvs-cache
agent.vcs_cache.enabled = true
agent.vcs_cache.path = /home/blissfox-server/.clearml/vcs-cache
agent.venv_update.enabled = false
agent.pip_download_cache.enabled = true
agent.pip_download_cache.path = /home/blissfox-server/.clearml/pip-download-cache
agent.translate_ssh = true
agent.reload_config = false
agent.docker_pip_cache = /home/blissfox-server/.clearml/pip-cache
agent.docker_apt_cache = /home/blissfox-server/.clearml/apt-cache
agent.docker_force_pull = false
agent.default_docker.image = nvidia/cuda:11.0.3-cudnn8-runtime-ubuntu20.04
agent.default_docker.match_rules.0.image = python:3.6-bullseye
agent.default_docker.match_rules.0.arguments = --ipc\=host
agent.default_docker.match_rules.0.match.script.binary = python3.6$
agent.default_docker.match_rules.1.image = python:3.7-bullseye
agent.default_docker.match_rules.1.arguments = --ipc\=host
agent.default_docker.match_rules.1.match.script.binary = python3.7$
agent.default_docker.match_rules.2.image = python:3.8-bullseye
agent.default_docker.match_rules.2.arguments = --ipc\=host
agent.default_docker.match_rules.2.match.script.binary = python3.8$
agent.default_docker.match_rules.3.image = python:3.9-bullseye
agent.default_docker.match_rules.3.arguments = --ipc\=host
agent.default_docker.match_rules.3.match.script.binary = python3.9$
agent.default_docker.match_rules.4.image = python:3.10-bullseye
agent.default_docker.match_rules.4.arguments = --ipc\=host
agent.default_docker.match_rules.4.match.script.binary = python3.10$
agent.default_docker.match_rules.5.image = python:3.11-bullseye
agent.default_docker.match_rules.5.arguments = --ipc\=host
agent.default_docker.match_rules.5.match.script.binary = python3.11$
agent.default_docker.match_rules.6.image = python:3.12-bullseye
agent.default_docker.match_rules.6.arguments = --ipc\=host
agent.default_docker.match_rules.6.match.script.binary = python3.12$
agent.enable_task_env = false
agent.sanitize_config_printout = ****
agent.hide_docker_command_env_vars.enabled = true
agent.hide_docker_command_env_vars.parse_embedded_urls = true
agent.abort_callback_max_timeout = 1800
agent.docker_internal_mounts.sdk_cache = /clearml_agent_cache
agent.docker_internal_mounts.apt_cache = /var/cache/apt/archives
agent.docker_internal_mounts.ssh_folder = ~/.ssh
agent.docker_internal_mounts.ssh_ro_folder = /.ssh
agent.docker_internal_mounts.pip_cache = /root/.cache/pip
agent.docker_internal_mounts.poetry_cache = /root/.cache/pypoetry
agent.docker_internal_mounts.vcs_cache = /root/.clearml/vcs-cache
agent.docker_internal_mounts.venvs_cache = /root/.clearml/venvs-cache
agent.docker_internal_mounts.venv_build = ~/.clearml/venvs-builds
agent.docker_internal_mounts.pip_download = /root/.clearml/pip-download-cache
agent.apply_environment = true
agent.apply_files = true
agent.custom_build_script = 
agent.disable_task_docker_override = false
agent.git_user = RR5555
agent.git_pass = ****
agent.default_python = 3.10
agent.cuda_version = 122
agent.cudnn_version = 0
sdk.storage.cache.default_base_dir = ~/.clearml/cache
sdk.storage.cache.size.min_free_bytes = 10GB
sdk.storage.direct_access.0.url = file://*
sdk.metrics.file_history_size = 100
sdk.metrics.matplotlib_untitled_history_size = 100
sdk.metrics.images.format = JPEG
sdk.metrics.images.quality = 87
sdk.metrics.images.subsampling = 0
sdk.metrics.tensorboard_single_series_per_graph = false
sdk.network.metrics.file_upload_threads = 4
sdk.network.metrics.file_upload_starvation_warning_sec = 120
sdk.network.iteration.max_retries_on_server_error = 5
sdk.network.iteration.retry_backoff_factor_sec = 10
sdk.network.file_upload_retries = 3
sdk.aws.s3.key = 
sdk.aws.s3.secret = ****
sdk.aws.s3.region = 
sdk.aws.s3.use_credentials_chain = false
sdk.aws.boto3.pool_connections = 512
sdk.aws.boto3.max_multipart_concurrency = 16
sdk.aws.boto3.multipart_threshold = 8388608
sdk.aws.boto3.multipart_chunksize = 8388608
sdk.log.null_log_propagate = false
sdk.log.task_log_buffer_capacity = 66
sdk.log.disable_urllib3_info = true
sdk.development.task_reuse_time_window_in_hours = 72.0
sdk.development.vcs_repo_detect_async = true
sdk.development.store_uncommitted_code_diff = true
sdk.development.support_stopping = true
sdk.development.default_output_uri = 
sdk.development.force_analyze_entire_repo = false
sdk.development.suppress_update_message = false
sdk.development.detect_with_pip_freeze = false
sdk.development.worker.report_period_sec = 2
sdk.development.worker.ping_period_sec = 30
sdk.development.worker.log_stdout = true
sdk.development.worker.report_global_mem_used = false
sdk.development.worker.report_event_flush_threshold = 100
sdk.development.worker.console_cr_flush_period = 10
sdk.apply_environment = false
sdk.apply_files = false
api.version = 1.5
api.verify_certificate = true
api.default_version = 1.5
api.http.max_req_size = 15728640
api.http.retries.total = 240
api.http.retries.connect = 240
api.http.retries.read = 240
api.http.retries.redirect = 240
api.http.retries.status = 240
api.http.retries.backoff_factor = 1.0
api.http.retries.backoff_max = 120.0
api.http.wait_on_maintenance_forever = true
api.http.pool_maxsize = 512
api.http.pool_connections = 512
api.auth.token_expiration_threshold_sec = ****
api.api_server = https://api.clearml.blissfox.xyz
api.web_server = https://app.clearml.blissfox.xyz
api.files_server = https://files.clearml.blissfox.xyz
api.credentials.access_key = EUAQEJNSXKQU1UN76PMC
api.credentials.secret_key = ****
api.host = https://api.clearml.blissfox.xyz

Executing task id [4f2118bf61c64795b89a3cd26b424fbe]:
repository = https://github.com/RR5555/clearml_demo.git
branch = main
version_num = 
tag = 
docker_cmd = rr5555/dith:cuda12.0-agent
entry_point = test.py
working_dir = .

created virtual environment CPython3.10.12.final.0-64 in 163ms
  creator CPython3Posix(dest=/home/blissfox-server/.clearml/venvs-builds/3.10, clear=False, no_vcs_ignore=False, global=False)
  seeder FromAppData(download=False, pip=bundle, setuptools=bundle, wheel=bundle, via=copy, app_data_dir=/home/blissfox-server/.local/share/virtualenv)
    added seed packages: pip==24.0, setuptools==70.0.0, wheel==0.43.0
  activators BashActivator,CShellActivator,FishActivator,NushellActivator,PowerShellActivator,PythonActivator


Using cached repository in "/home/blissfox-server/.clearml/vcs-cache/clearml_demo.git.3ae9b88477bc6079c4900a526f62cad3/clearml_demo.git"
pulling git
From https://github.com/RR5555/clearml_demo
   22c333e..1619339  main       -> origin/main
pulling git completed
Note: switching to 'origin/main'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by switching back to a branch.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -c with the switch command. Example:

  git switch -c <new-branch-name>

Or undo this operation with:

  git switch -

Turn off this advice by setting config variable advice.detachedHead to false

HEAD is now at 1619339 Continued client side: PAT & private repo
type: git
url: https://github.com/RR5555/clearml_demo.git
branch: HEAD
commit: 16193394d0c48ca7fe11e8d5a951b45d4c7ac7a5
root: /home/blissfox-server/.clearml/venvs-builds/3.10/task_repository/clearml_demo.git
Applying uncommitted changes


Ignoring pip: markers 'python_version < "3.10"' don't match your environment
Ignoring pip: markers 'python_version >= "3.12"' don't match your environment
Collecting pip<22.3
  Using cached pip-22.2.2-py3-none-any.whl.metadata (4.2 kB)
Using cached pip-22.2.2-py3-none-any.whl (2.0 MB)
Installing collected packages: pip
  Attempting uninstall: pip
    Found existing installation: pip 24.0
    Uninstalling pip-24.0:
      Successfully uninstalled pip-24.0
Successfully installed pip-22.2.2
Collecting Cython
  Using cached Cython-3.0.11-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.6 MB)
Installing collected packages: Cython
Successfully installed Cython-3.0.11
Trying pip install: /home/blissfox-server/.clearml/venvs-builds/3.10/task_repository/clearml_demo.git/requirements.txt
Collecting numpy
  Using cached numpy-2.1.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (16.3 MB)
Installing collected packages: numpy
Successfully installed numpy-2.1.1
Torch CUDA 121 index page found, adding `https://download.pytorch.org/whl/cu121/`
PyTorch: Adding index `https://download.pytorch.org/whl/cu121/` and installing `torch`
Torch CUDA 121 index page found, adding `https://download.pytorch.org/whl/cu121/`
PyTorch: Adding index `https://download.pytorch.org/whl/cu121/` and installing `torchvision`
Looking in indexes: https://pypi.org/simple, https://download.pytorch.org/whl/cu121/
Collecting clearml-agent
  Using cached clearml_agent-1.9.1-py3-none-any.whl (437 kB)
Collecting scikit-learn
  Using cached scikit_learn-1.5.2-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (13.3 MB)
Collecting pandas
  Using cached pandas-2.2.3-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (13.1 MB)
Collecting joblib
  Using cached joblib-1.4.2-py3-none-any.whl (301 kB)
Collecting matplotlib
  Using cached matplotlib-3.9.2-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (8.3 MB)
Requirement already satisfied: numpy in /home/blissfox-server/.clearml/venvs-builds/3.10/lib/python3.10/site-packages (from -r /tmp/requirements_ekr9gq27.txt (line 8)) (2.1.1)
Collecting torch
  Using cached https://download.pytorch.org/whl/cu121/torch-2.4.1%2Bcu121-cp310-cp310-linux_x86_64.whl (798.9 MB)
Collecting torchvision
  Using cached https://download.pytorch.org/whl/cu121/torchvision-0.19.1%2Bcu121-cp310-cp310-linux_x86_64.whl (7.1 MB)
Collecting tensorboard
  Using cached tensorboard-2.18.0-py3-none-any.whl (5.5 MB)
Collecting ipywidgets
  Using cached ipywidgets-8.1.5-py3-none-any.whl (139 kB)
Collecting clearml
  Using cached clearml-1.16.4-py2.py3-none-any.whl (1.2 MB)
Collecting pathlib2<2.4.0,>=2.3.0
  Using cached pathlib2-2.3.7.post1-py2.py3-none-any.whl (18 kB)
Collecting PyYAML<6.1,>=3.12
  Using cached PyYAML-6.0.2-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (751 kB)
Collecting urllib3<2,>=1.21.1
  Using cached urllib3-1.26.20-py2.py3-none-any.whl (144 kB)
Collecting virtualenv<21,>=16
  Downloading virtualenv-20.26.6-py3-none-any.whl (6.0 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.0/6.0 MB 18.1 MB/s eta 0:00:00
Collecting jsonschema<5.0.0,>=2.6.0
  Using cached jsonschema-4.23.0-py3-none-any.whl (88 kB)
Collecting attrs<24.0.0,>=18.0
  Using cached attrs-23.2.0-py3-none-any.whl (60 kB)
Collecting python-dateutil<2.9.0,>=2.4.2
  Using cached python_dateutil-2.8.2-py2.py3-none-any.whl (247 kB)
Collecting pyparsing<3.2.0,>=2.0.3
  Using cached pyparsing-3.1.4-py3-none-any.whl (104 kB)
Collecting requests<=2.31.0,>=2.20.0
  Using cached requests-2.31.0-py3-none-any.whl (62 kB)
Collecting pyjwt<2.9.0,>=2.4.0
  Using cached PyJWT-2.8.0-py3-none-any.whl (22 kB)
Collecting psutil<5.10.0,>=3.4.2
  Using cached psutil-5.9.8-cp36-abi3-manylinux_2_12_x86_64.manylinux2010_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl (288 kB)
Collecting furl<2.2.0,>=2.0.0
  Using cached furl-2.1.3-py2.py3-none-any.whl (20 kB)
Collecting six<1.17.0,>=1.13.0
  Using cached six-1.16.0-py2.py3-none-any.whl (11 kB)
Collecting scipy>=1.6.0
  Using cached scipy-1.14.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (41.2 MB)
Collecting threadpoolctl>=3.1.0
  Using cached threadpoolctl-3.5.0-py3-none-any.whl (18 kB)
Collecting tzdata>=2022.7
  Using cached tzdata-2024.2-py2.py3-none-any.whl (346 kB)
Collecting pytz>=2020.1
  Using cached pytz-2024.2-py2.py3-none-any.whl (508 kB)
Collecting fonttools>=4.22.0
  Using cached fonttools-4.54.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (4.6 MB)
Collecting kiwisolver>=1.3.1
  Using cached kiwisolver-1.4.7-cp310-cp310-manylinux_2_12_x86_64.manylinux2010_x86_64.whl (1.6 MB)
Collecting packaging>=20.0
  Using cached packaging-24.1-py3-none-any.whl (53 kB)
Collecting contourpy>=1.0.1
  Using cached contourpy-1.3.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (322 kB)
Collecting cycler>=0.10
  Using cached cycler-0.12.1-py3-none-any.whl (8.3 kB)
Collecting pillow>=8
  Using cached pillow-10.4.0-cp310-cp310-manylinux_2_28_x86_64.whl (4.5 MB)
Collecting nvidia-cusparse-cu12==12.1.0.106
  Using cached https://download.pytorch.org/whl/cu121/nvidia_cusparse_cu12-12.1.0.106-py3-none-manylinux1_x86_64.whl (196.0 MB)
Collecting nvidia-nvtx-cu12==12.1.105
  Using cached https://download.pytorch.org/whl/cu121/nvidia_nvtx_cu12-12.1.105-py3-none-manylinux1_x86_64.whl (99 kB)
Collecting nvidia-cusolver-cu12==11.4.5.107
  Using cached https://download.pytorch.org/whl/cu121/nvidia_cusolver_cu12-11.4.5.107-py3-none-manylinux1_x86_64.whl (124.2 MB)
Collecting jinja2
  Using cached jinja2-3.1.4-py3-none-any.whl (133 kB)
Collecting nvidia-cuda-runtime-cu12==12.1.105
  Using cached https://download.pytorch.org/whl/cu121/nvidia_cuda_runtime_cu12-12.1.105-py3-none-manylinux1_x86_64.whl (823 kB)
Collecting nvidia-cuda-cupti-cu12==12.1.105
  Using cached https://download.pytorch.org/whl/cu121/nvidia_cuda_cupti_cu12-12.1.105-py3-none-manylinux1_x86_64.whl (14.1 MB)
Collecting nvidia-cufft-cu12==11.0.2.54
  Using cached https://download.pytorch.org/whl/cu121/nvidia_cufft_cu12-11.0.2.54-py3-none-manylinux1_x86_64.whl (121.6 MB)
Collecting nvidia-cudnn-cu12==9.1.0.70
  Using cached https://download.pytorch.org/whl/cu121/nvidia_cudnn_cu12-9.1.0.70-py3-none-manylinux2014_x86_64.whl (664.8 MB)
Collecting triton==3.0.0
  Using cached https://download.pytorch.org/whl/triton-3.0.0-1-cp310-cp310-manylinux2014_x86_64.manylinux_2_17_x86_64.whl (209.4 MB)
Collecting fsspec
  Using cached fsspec-2024.9.0-py3-none-any.whl (179 kB)
Collecting filelock
  Using cached filelock-3.16.1-py3-none-any.whl (16 kB)
Collecting typing-extensions>=4.8.0
  Using cached typing_extensions-4.12.2-py3-none-any.whl (37 kB)
Collecting networkx
  Using cached networkx-3.3-py3-none-any.whl (1.7 MB)
Collecting nvidia-cublas-cu12==12.1.3.1
  Using cached https://download.pytorch.org/whl/cu121/nvidia_cublas_cu12-12.1.3.1-py3-none-manylinux1_x86_64.whl (410.6 MB)
Collecting nvidia-cuda-nvrtc-cu12==12.1.105
  Using cached https://download.pytorch.org/whl/cu121/nvidia_cuda_nvrtc_cu12-12.1.105-py3-none-manylinux1_x86_64.whl (23.7 MB)
Collecting nvidia-curand-cu12==10.3.2.106
  Using cached https://download.pytorch.org/whl/cu121/nvidia_curand_cu12-10.3.2.106-py3-none-manylinux1_x86_64.whl (56.5 MB)
Collecting sympy
  Using cached sympy-1.13.3-py3-none-any.whl (6.2 MB)
Collecting nvidia-nccl-cu12==2.20.5
  Using cached https://download.pytorch.org/whl/cu121/nvidia_nccl_cu12-2.20.5-py3-none-manylinux2014_x86_64.whl (176.2 MB)
Collecting nvidia-nvjitlink-cu12
  Using cached nvidia_nvjitlink_cu12-12.6.68-py3-none-manylinux2014_x86_64.whl (19.7 MB)
Collecting grpcio>=1.48.2
  Using cached grpcio-1.66.2-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (5.8 MB)
Collecting tensorboard-data-server<0.8.0,>=0.7.0
  Using cached tensorboard_data_server-0.7.2-py3-none-manylinux_2_31_x86_64.whl (6.6 MB)
Collecting markdown>=2.6.8
  Using cached Markdown-3.7-py3-none-any.whl (106 kB)
Collecting absl-py>=0.4
  Using cached absl_py-2.1.0-py3-none-any.whl (133 kB)
Collecting protobuf!=4.24.0,>=3.19.6
  Using cached protobuf-5.28.2-cp38-abi3-manylinux2014_x86_64.whl (316 kB)
Requirement already satisfied: setuptools>=41.0.0 in /home/blissfox-server/.clearml/venvs-builds/3.10/lib/python3.10/site-packages (from tensorboard->-r /tmp/requirements_ekr9gq27.txt (line 11)) (70.0.0)
Collecting werkzeug>=1.0.1
  Using cached werkzeug-3.0.4-py3-none-any.whl (227 kB)
Collecting comm>=0.1.3
  Using cached comm-0.2.2-py3-none-any.whl (7.2 kB)
Collecting traitlets>=4.3.1
  Using cached traitlets-5.14.3-py3-none-any.whl (85 kB)
Collecting widgetsnbextension~=4.0.12
  Using cached widgetsnbextension-4.0.13-py3-none-any.whl (2.3 MB)
Collecting ipython>=6.1.0
  Using cached ipython-8.27.0-py3-none-any.whl (818 kB)
Collecting jupyterlab-widgets~=3.0.12
  Using cached jupyterlab_widgets-3.0.13-py3-none-any.whl (214 kB)
Collecting referencing<0.40
  Using cached referencing-0.35.1-py3-none-any.whl (26 kB)
Collecting orderedmultidict>=1.0.1
  Using cached orderedmultidict-1.0.1-py2.py3-none-any.whl (11 kB)
Collecting pygments>=2.4.0
  Using cached pygments-2.18.0-py3-none-any.whl (1.2 MB)
Collecting pexpect>4.3
  Using cached pexpect-4.9.0-py2.py3-none-any.whl (63 kB)
Collecting stack-data
  Using cached stack_data-0.6.3-py3-none-any.whl (24 kB)
Collecting prompt-toolkit<3.1.0,>=3.0.41
  Downloading prompt_toolkit-3.0.48-py3-none-any.whl (386 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 386.6/386.6 kB 7.6 MB/s eta 0:00:00
Collecting matplotlib-inline
  Using cached matplotlib_inline-0.1.7-py3-none-any.whl (9.9 kB)
Collecting decorator
  Using cached decorator-5.1.1-py3-none-any.whl (9.1 kB)
Collecting jedi>=0.16
  Using cached jedi-0.19.1-py2.py3-none-any.whl (1.6 MB)
Collecting exceptiongroup
  Using cached exceptiongroup-1.2.2-py3-none-any.whl (16 kB)
Collecting rpds-py>=0.7.1
  Using cached rpds_py-0.20.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (354 kB)
Collecting jsonschema-specifications>=2023.03.6
  Using cached jsonschema_specifications-2023.12.1-py3-none-any.whl (18 kB)
Collecting certifi>=2017.4.17
  Using cached certifi-2024.8.30-py3-none-any.whl (167 kB)
Collecting idna<4,>=2.5
  Using cached idna-3.10-py3-none-any.whl (70 kB)
Collecting charset-normalizer<4,>=2
  Using cached charset_normalizer-3.3.2-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (142 kB)
Collecting platformdirs<5,>=3.9.1
  Downloading platformdirs-4.3.6-py3-none-any.whl (18 kB)
Collecting distlib<1,>=0.3.7
  Using cached distlib-0.3.8-py2.py3-none-any.whl (468 kB)
Collecting MarkupSafe>=2.1.1
  Using cached https://download.pytorch.org/whl/MarkupSafe-2.1.5-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (25 kB)
Collecting mpmath<1.4,>=1.1.0
  Using cached https://download.pytorch.org/whl/mpmath-1.3.0-py3-none-any.whl (536 kB)
Collecting parso<0.9.0,>=0.8.3
  Using cached parso-0.8.4-py2.py3-none-any.whl (103 kB)
Collecting ptyprocess>=0.5
  Using cached ptyprocess-0.7.0-py2.py3-none-any.whl (13 kB)
Collecting wcwidth
  Using cached wcwidth-0.2.13-py2.py3-none-any.whl (34 kB)
Collecting executing>=1.2.0
  Using cached executing-2.1.0-py2.py3-none-any.whl (25 kB)
Collecting pure-eval
  Using cached pure_eval-0.2.3-py3-none-any.whl (11 kB)
Collecting asttokens>=2.1.0
  Using cached asttokens-2.4.1-py2.py3-none-any.whl (27 kB)
Installing collected packages: wcwidth, pytz, pure-eval, ptyprocess, mpmath, distlib, widgetsnbextension, urllib3, tzdata, typing-extensions, traitlets, threadpoolctl, tensorboard-data-server, sympy, six, scipy, rpds-py, PyYAML, pyparsing, pyjwt, pygments, psutil, protobuf, prompt-toolkit, platformdirs, pillow, pexpect, parso, packaging, nvidia-nvtx-cu12, nvidia-nvjitlink-cu12, nvidia-nccl-cu12, nvidia-curand-cu12, nvidia-cufft-cu12, nvidia-cuda-runtime-cu12, nvidia-cuda-nvrtc-cu12, nvidia-cuda-cupti-cu12, nvidia-cublas-cu12, networkx, MarkupSafe, markdown, kiwisolver, jupyterlab-widgets, joblib, idna, grpcio, fsspec, fonttools, filelock, executing, exceptiongroup, decorator, cycler, contourpy, charset-normalizer, certifi, attrs, absl-py, werkzeug, virtualenv, triton, scikit-learn, requests, referencing, python-dateutil, pathlib2, orderedmultidict, nvidia-cusparse-cu12, nvidia-cudnn-cu12, matplotlib-inline, jinja2, jedi, comm, asttokens, tensorboard, stack-data, pandas, nvidia-cusolver-cu12, matplotlib, jsonschema-specifications, furl, torch, jsonschema, ipython, torchvision, ipywidgets, clearml-agent, clearml
Successfully installed MarkupSafe-2.1.5 PyYAML-6.0.2 absl-py-2.1.0 asttokens-2.4.1 attrs-23.2.0 certifi-2024.8.30 charset-normalizer-3.3.2 clearml-1.16.4 clearml-agent-1.9.1 comm-0.2.2 contourpy-1.3.0 cycler-0.12.1 decorator-5.1.1 distlib-0.3.8 exceptiongroup-1.2.2 executing-2.1.0 filelock-3.16.1 fonttools-4.54.1 fsspec-2024.9.0 furl-2.1.3 grpcio-1.66.2 idna-3.10 ipython-8.27.0 ipywidgets-8.1.5 jedi-0.19.1 jinja2-3.1.4 joblib-1.4.2 jsonschema-4.23.0 jsonschema-specifications-2023.12.1 jupyterlab-widgets-3.0.13 kiwisolver-1.4.7 markdown-3.7 matplotlib-3.9.2 matplotlib-inline-0.1.7 mpmath-1.3.0 networkx-3.3 nvidia-cublas-cu12-12.1.3.1 nvidia-cuda-cupti-cu12-12.1.105 nvidia-cuda-nvrtc-cu12-12.1.105 nvidia-cuda-runtime-cu12-12.1.105 nvidia-cudnn-cu12-9.1.0.70 nvidia-cufft-cu12-11.0.2.54 nvidia-curand-cu12-10.3.2.106 nvidia-cusolver-cu12-11.4.5.107 nvidia-cusparse-cu12-12.1.0.106 nvidia-nccl-cu12-2.20.5 nvidia-nvjitlink-cu12-12.6.68 nvidia-nvtx-cu12-12.1.105 orderedmultidict-1.0.1 packaging-24.1 pandas-2.2.3 parso-0.8.4 pathlib2-2.3.7.post1 pexpect-4.9.0 pillow-10.4.0 platformdirs-4.3.6 prompt-toolkit-3.0.48 protobuf-5.28.2 psutil-5.9.8 ptyprocess-0.7.0 pure-eval-0.2.3 pygments-2.18.0 pyjwt-2.8.0 pyparsing-3.1.4 python-dateutil-2.8.2 pytz-2024.2 referencing-0.35.1 requests-2.31.0 rpds-py-0.20.0 scikit-learn-1.5.2 scipy-1.14.1 six-1.16.0 stack-data-0.6.3 sympy-1.13.3 tensorboard-2.18.0 tensorboard-data-server-0.7.2 threadpoolctl-3.5.0 torch-2.4.1+cu121 torchvision-0.19.1+cu121 traitlets-5.14.3 triton-3.0.0 typing-extensions-4.12.2 tzdata-2024.2 urllib3-1.26.20 virtualenv-20.26.6 wcwidth-0.2.13 werkzeug-3.0.4 widgetsnbextension-4.0.13
Adding venv into cache: /home/blissfox-server/.clearml/venvs-builds/3.10
Running task id [4f2118bf61c64795b89a3cd26b424fbe]:
[.]$ /home/blissfox-server/.clearml/venvs-builds/3.10/bin/python -u test.py
Summary - installed python packages:
pip:
- absl-py==2.1.0
- asttokens==2.4.1
- attrs==23.2.0
- certifi==2024.8.30
- charset-normalizer==3.3.2
- clearml==1.16.4
- comm==0.2.2
- contourpy==1.3.0
- cycler==0.12.1
- Cython==3.0.11
- decorator==5.1.1
- distlib==0.3.8
- exceptiongroup==1.2.2
- executing==2.1.0
- filelock==3.16.1
- fonttools==4.54.1
- fsspec==2024.9.0
- furl==2.1.3
- grpcio==1.66.2
- idna==3.10
- ipython==8.27.0
- ipywidgets==8.1.5
- jedi==0.19.1
- Jinja2==3.1.4
- joblib==1.4.2
- jsonschema==4.23.0
- jsonschema-specifications==2023.12.1
- jupyterlab_widgets==3.0.13
- kiwisolver==1.4.7
- Markdown==3.7
- MarkupSafe==2.1.5
- matplotlib==3.9.2
- matplotlib-inline==0.1.7
- mpmath==1.3.0
- networkx==3.3
- numpy==2.1.1
- nvidia-cublas-cu12==12.1.3.1
- nvidia-cuda-cupti-cu12==12.1.105
- nvidia-cuda-nvrtc-cu12==12.1.105
- nvidia-cuda-runtime-cu12==12.1.105
- nvidia-cudnn-cu12==9.1.0.70
- nvidia-cufft-cu12==11.0.2.54
- nvidia-curand-cu12==10.3.2.106
- nvidia-cusolver-cu12==11.4.5.107
- nvidia-cusparse-cu12==12.1.0.106
- nvidia-nccl-cu12==2.20.5
- nvidia-nvjitlink-cu12==12.6.68
- nvidia-nvtx-cu12==12.1.105
- orderedmultidict==1.0.1
- packaging==24.1
- pandas==2.2.3
- parso==0.8.4
- pathlib2==2.3.7.post1
- pexpect==4.9.0
- pillow==10.4.0
- platformdirs==4.3.6
- prompt_toolkit==3.0.48
- protobuf==5.28.2
- psutil==5.9.8
- ptyprocess==0.7.0
- pure_eval==0.2.3
- Pygments==2.18.0
- PyJWT==2.8.0
- pyparsing==3.1.4
- python-dateutil==2.8.2
- pytz==2024.2
- PyYAML==6.0.2
- referencing==0.35.1
- requests==2.31.0
- rpds-py==0.20.0
- scikit-learn==1.5.2
- scipy==1.14.1
- six==1.16.0
- stack-data==0.6.3
- sympy==1.13.3
- tensorboard==2.18.0
- tensorboard-data-server==0.7.2
- threadpoolctl==3.5.0
- torch==2.4.1+cu121
- torchvision==0.19.1+cu121
- traitlets==5.14.3
- triton==3.0.0
- typing_extensions==4.12.2
- tzdata==2024.2
- urllib3==1.26.20
- virtualenv==20.26.6
- wcwidth==0.2.13
- Werkzeug==3.0.4
- widgetsnbextension==4.0.13

Environment setup completed successfully

Starting Task Execution:

ClearML results page: https://app.clearml.blissfox.xyz/projects/ce6d125a264149bc89e8e62fb1dd0d7a/experiments/4f2118bf61c64795b89a3cd26b424fbe/output/log
3.10.12 (main, Sep 11 2024, 15:47:36) [GCC 11.4.0]
['/home/blissfox-server/.clearml/venvs-builds/3.10/task_repository/clearml_demo.git', '/home/blissfox-server/.clearml/venvs-builds/3.10/task_repository/clearml_demo.git', '/usr/lib/python310.zip', '/usr/lib/python3.10', '/usr/lib/python3.10/lib-dynload', '/home/blissfox-server/.clearml/venvs-builds/3.10/lib/python3.10/site-packages']
2024-09-30 01:03:47,164 [INFO] Started
2024-09-30 01:03:47,476 [INFO] type of a:<class 'torch.Tensor'>
2024-09-30 01:04:07,513 [INFO] Finished

Leaving process id 187795
DONE: Running task '4f2118bf61c64795b89a3cd26b424fbe', exit status 0
Process completed successfully
```

</details>

The new output is composed of a new header:
```bash
task 4f2118bf61c64795b89a3cd26b424fbe pulled from 6b55ab8e3be4482eb53acfed3e4037a5 by worker clearml-agent-bengio:gpu0
Running task '4f2118bf61c64795b89a3cd26b424fbe'
Storing stdout and stderr log to '/tmp/.clearml_agent_out.np498ozg.txt', '/tmp/.clearml_agent_out.np498ozg.txt'
Current configuration (clearml_agent v1.9.1, location: /tmp/.clearml_agent.3tikdf0t.cfg):
```
the config. in the tmp config file(`/tmp/.clearml_agent.3tikdf0t.cfg`),\
AND the console output of the new task!

> Note that the supposed output file (`/tmp/.clearml_agent_out.np498ozg.txt`) does not exist.\
> TODO: Maybe it is actually located on the server side?

In [21]:
%%bash
# cat /tmp/.clearml_agent_out.np498ozg.txt
ls /tmp/.clearml*

/tmp/.clearmlagent_167497_9u1qoi1a.tmp
/tmp/.clearml_agent.3tikdf0t.cfg
/tmp/.clearml_agent_daemon_outho7nopyx.txt


In [ ]:
%%bash
cat /tmp/.clearml_agent.*.cfg

<details>
<summary>Output</summary>

```bash
{
    "agent": {
        "worker_id": "clearml-agent-bengio:gpu0",
        "worker_name": "clearml-agent-bengio",
        "force_git_ssh_protocol": false,
        "python_binary": "/usr/bin/python3.10",
        "package_manager": {
            "type": "pip",
            "pip_version": [
                "<20.2 ; python_version < '3.10'",
                "<22.3 ; python_version >= '3.10' and python_version <= '3.11'",
                ">=23,<24.3 ; python_version >= '3.12'"
            ],
            "system_site_packages": false,
            "force_upgrade": false,
            "conda_channels": [
                "pytorch",
                "conda-forge",
                "nvidia",
                "defaults"
            ],
            "priority_optional_packages": [
                "pygobject"
            ],
            "torch_nightly": false,
            "poetry_files_from_repo_working_dir": false,
            "extra_index_url": []
        },
        "venvs_dir": "/home/blissfox-server/.clearml/venvs-builds",
        "venvs_cache": {
            "max_entries": 10,
            "free_space_threshold_gb": 2.0,
            "path": "~/.clearml/venvs-cache"
        },
        "vcs_cache": {
            "enabled": true,
            "path": "/home/blissfox-server/.clearml/vcs-cache"
        },
        "venv_update": {
            "enabled": false
        },
        "pip_download_cache": {
            "enabled": true,
            "path": "/home/blissfox-server/.clearml/pip-download-cache"
        },
        "translate_ssh": true,
        "reload_config": false,
        "docker_pip_cache": "/home/blissfox-server/.clearml/pip-cache",
        "docker_apt_cache": "/home/blissfox-server/.clearml/apt-cache",
        "docker_force_pull": false,
        "default_docker": {
            "image": "nvidia/cuda:11.0.3-cudnn8-runtime-ubuntu20.04",
            "match_rules": [
                {
                    "image": "python:3.6-bullseye",
                    "arguments": "--ipc=host",
                    "match": {
                        "script": {
                            "binary": "python3.6$"
                        }
                    }
                },
                {
                    "image": "python:3.7-bullseye",
                    "arguments": "--ipc=host",
                    "match": {
                        "script": {
                            "binary": "python3.7$"
                        }
                    }
                },
                {
                    "image": "python:3.8-bullseye",
                    "arguments": "--ipc=host",
                    "match": {
                        "script": {
                            "binary": "python3.8$"
                        }
                    }
                },
                {
                    "image": "python:3.9-bullseye",
                    "arguments": "--ipc=host",
                    "match": {
                        "script": {
                            "binary": "python3.9$"
                        }
                    }
                },
                {
                    "image": "python:3.10-bullseye",
                    "arguments": "--ipc=host",
                    "match": {
                        "script": {
                            "binary": "python3.10$"
                        }
                    }
                },
                {
                    "image": "python:3.11-bullseye",
                    "arguments": "--ipc=host",
                    "match": {
                        "script": {
                            "binary": "python3.11$"
                        }
                    }
                },
                {
                    "image": "python:3.12-bullseye",
                    "arguments": "--ipc=host",
                    "match": {
                        "script": {
                            "binary": "python3.12$"
                        }
                    }
                }
            ]
        },
        "enable_task_env": false,
        "sanitize_config_printout": {
            "hide_secrets": [
                "^sanitize_config_printout$",
                "secret",
                "pass",
                "token",
                "account_key",
                "contents"
            ],
            "hide_secrets_recursive": [
                "^environment$"
            ],
            "dont_hide_secrets": [
                "^enable_git_ask_pass$"
            ],
            "docker_commands": [
                "^extra_docker_arguments$"
            ],
            "urls": [
                "^extra_index_url$"
            ]
        },
        "hide_docker_command_env_vars": {
            "enabled": true,
            "extra_keys": [],
            "parse_embedded_urls": true
        },
        "abort_callback_max_timeout": 1800,
        "docker_internal_mounts": {
            "sdk_cache": "/clearml_agent_cache",
            "apt_cache": "/var/cache/apt/archives",
            "ssh_folder": "~/.ssh",
            "ssh_ro_folder": "/.ssh",
            "pip_cache": "/root/.cache/pip",
            "poetry_cache": "/root/.cache/pypoetry",
            "vcs_cache": "/root/.clearml/vcs-cache",
            "venvs_cache": "/root/.clearml/venvs-cache",
            "venv_build": "~/.clearml/venvs-builds",
            "pip_download": "/root/.clearml/pip-download-cache"
        },
        "apply_environment": true,
        "apply_files": true,
        "custom_build_script": "",
        "disable_task_docker_override": false,
        "git_user": "RR5555",
        "git_pass": "github_pat_**********************_***********************************************************",
        "default_python": "3.10",
        "cuda_version": "122",
        "cudnn_version": "0"
    },
    "sdk": {
        "storage": {
            "cache": {
                "default_base_dir": "~/.clearml/cache",
                "size": {
                    "min_free_bytes": "10GB"
                }
            },
            "direct_access": [
                {
                    "url": "file://*"
                }
            ]
        },
        "metrics": {
            "file_history_size": 100,
            "matplotlib_untitled_history_size": 100,
            "images": {
                "format": "JPEG",
                "quality": 87,
                "subsampling": 0
            },
            "tensorboard_single_series_per_graph": false
        },
        "network": {
            "metrics": {
                "file_upload_threads": 4,
                "file_upload_starvation_warning_sec": 120
            },
            "iteration": {
                "max_retries_on_server_error": 5,
                "retry_backoff_factor_sec": 10
            },
            "file_upload_retries": 3
        },
        "aws": {
            "s3": {
                "key": "",
                "secret": "",
                "region": "",
                "credentials": [],
                "use_credentials_chain": false,
                "extra_args": {}
            },
            "boto3": {
                "pool_connections": 512,
                "max_multipart_concurrency": 16,
                "multipart_threshold": 8388608,
                "multipart_chunksize": 8388608
            }
        },
        "google": {
            "storage": {}
        },
        "azure": {
            "storage": {}
        },
        "log": {
            "null_log_propagate": false,
            "task_log_buffer_capacity": 66,
            "disable_urllib3_info": true
        },
        "development": {
            "task_reuse_time_window_in_hours": 72.0,
            "vcs_repo_detect_async": true,
            "store_uncommitted_code_diff": true,
            "support_stopping": true,
            "default_output_uri": "",
            "force_analyze_entire_repo": false,
            "suppress_update_message": false,
            "detect_with_pip_freeze": false,
            "worker": {
                "report_period_sec": 2,
                "ping_period_sec": 30,
                "log_stdout": true,
                "report_global_mem_used": false,
                "report_event_flush_threshold": 100,
                "console_cr_flush_period": 10
            },
            "log_os_environments": []
        },
        "apply_environment": false,
        "apply_files": false
    },
    "logging": {
        "version": 1,
        "loggers": {
            "urllib3": {
                "level": "ERROR"
            }
        }
    },
    "api": {
        "version": 1.5,
        "verify_certificate": true,
        "default_version": 1.5,
        "http": {
            "max_req_size": 15728640,
            "retries": {
                "total": 240,
                "connect": 240,
                "read": 240,
                "redirect": 240,
                "status": 240,
                "backoff_factor": 1.0,
                "backoff_max": 120.0
            },
            "wait_on_maintenance_forever": true,
            "pool_maxsize": 512,
            "pool_connections": 512
        },
        "auth": {
            "token_expiration_threshold_sec": 43200
        },
        "api_server": "https://api.clearml.blissfox.xyz",
        "web_server": "https://app.clearml.blissfox.xyz",
        "files_server": "https://files.clearml.blissfox.xyz",
        "credentials": {
            "access_key": "EUAQEJNSXKQU1UN76PMC",
            "secret_key": "**************************************************"
        },
        "host": "https://api.clearml.blissfox.xyz"
    },
    "env": "default"
}
```

</details>

I do not see any noticeable change in the cfg file.

In [24]:
%%bash
cat /tmp/.clearmlagent_*.tmp

clearml-agent-bengio:gpu0
0

<details>
<summary>Output</summary>

```bash
clearml-agent-bengio:gpu0
0
```

</details>

No change.

**Stop the agent:**

In [25]:
%%bash
clearml-agent daemon --stop clearml-agent-bengio:gpu0

Terminating clearml-agent worker_id=clearml-agent-bengio:gpu0 pid=167497


In [26]:
%%bash
clearml-agent list

workers: []



# Launch an exp. from a public repo.

We will:
* Launch an agent/worker on a brand new queue: it will wait for a load (created a container? Or container only under load?)
* Launch a task from the CLI calling a script from a public GitHub repo
* Stop the agent

In [ ]:
%%bash
CLEARML_WORKER_NAME=clearml-agent-bengio CLEARML_WORKER_ID=clearml-agent-bengio:gpu0 clearml-agent daemon --detached --queue gpu_dithering --create-queue --gpus 0
clearml-task --project Demo --name clearml_test --script test.py --repo https://github.com/RR5555/clearml_demo.git --branch main --queue gpu_dithering --docker rr5555/dith:cuda12.0-agent
clearml-agent daemon --stop clearml-agent-bengio:gpu0

# Check out on Tasks from Python



Unfortunately, it seems that, for now (2024/09/18), there is no CLI cmd to monitor the tasks.\
No such things as `clearml-task list`.

We thus have to fall back to the python interface: [ClearML -- Accessing Tasks](https://clear.ml/docs/latest/docs/clearml_sdk/task_sdk/#accessing-tasks)

[ClearML -- Task get_status](https://clear.ml/docs/latest/docs/references/sdk/task/#get_status):
* TaskStatusEnum: [“created”, “in_progress”, “stopped”, “closed”, “failed”, “completed”, “queued”, “published”, “publishing”, “unknown”]

In [75]:
from clearml import Task

> Do not worry about reruning the import cell above as, if the python module is already loaded, it won't be loaded twice.

Get all the tasks `in_progress` in the project `Demo`:

In [4]:
in_progress_task_list = Task.get_tasks(
    task_ids=None,  # type Optional[Sequence[str]]
    project_name='Demo',  # Optional[str]
    task_name=None,  # Optional[str]
    allow_archived=True, # [bool]
    task_filter={
        'status': ('in_progress',)
    },  # Optional[Dict]#
    # tasks with tag `included_tag` or without tag `excluded_tag`
    tags=None #['included_tag', '-excluded_tag'] 
)
print(f"Number of `in_progress` task: {len(in_progress_task_list)}")
print(in_progress_task_list)

[]


Get the tasks `completed` in project `Demo` ordered by `last_update`:

In [76]:
completed_task_list = Task.get_tasks(
    task_ids=None,  # type Optional[Sequence[str]]
    project_name='Demo',  # Optional[str]
    task_name=None,  # Optional[str]
    allow_archived=True, # [bool]
    task_filter={
        'status': ('completed',),
        'order_by': ('last_update',)
    },  # Optional[Dict]#
    # tasks with tag `included_tag` or without tag `excluded_tag`
    tags=None #['included_tag', '-excluded_tag'] 
)
print(f"Number of `completed` task: {len(completed_task_list)}")
print(completed_task_list)

Number of `completed` task: 13
[<clearml.task.Task object at 0x737a88172ad0>, <clearml.task.Task object at 0x737a88172b60>, <clearml.task.Task object at 0x737a7df57070>, <clearml.task.Task object at 0x737a7df578b0>, <clearml.task.Task object at 0x737a7df900d0>, <clearml.task.Task object at 0x737a7df90910>, <clearml.task.Task object at 0x737a7df91120>, <clearml.task.Task object at 0x737a7df918a0>, <clearml.task.Task object at 0x737a7df91ff0>, <clearml.task.Task object at 0x737a7df928c0>, <clearml.task.Task object at 0x737a7df93040>, <clearml.task.Task object at 0x737a7df937c0>, <clearml.task.Task object at 0x737a7dfc0190>]


## Explore a task INFO panel information

Access the information from the `INFO` panel of the clearml GUI from the last task in the previous list:

In [77]:
_task = completed_task_list[-1]

If you want to explore the fields of the task, just list them `obj.__dir__()` or `dir(obj)`:
```python
print(_task.__dir__())
print(tuple(_field for _field in _task.__dir__() if _field.find('get')>-1))
```

In [ ]:
# print(_task.__dir__())
# print(tuple(_field for _field in _task.__dir__() if _field.find('get')>-1))
print(tuple(_field for _field in _task.__dir__() if _field.find('name')>-1))

Let's look at some of the task informative properties:

In [8]:
from IPython.display import display, Markdown
from collections import OrderedDict

_cmds = OrderedDict((
            ('_task.get_reported_console_output()', _task.get_reported_console_output()),
            ('_task._get_runtime_properties()', _task._get_runtime_properties()),
            ('_task.get_tags()', _task.get_tags()),
            ('_task.get_script()', _task.get_script()),
            ('_task.comment', _task.comment),
            ('_task.status', _task.status),
            ('_task.get_parameters()', _task.get_parameters()),
            ('_task.get_task_output_log_web_page(_task.id)', _task.get_task_output_log_web_page(_task.id)),
            ('_task.name', _task.name),
            ('_task.id', _task.id),
            ('_task._project_name', _task._project_name),
            ('_task.get_output_destination()', _task.get_output_destination())
))

_md_table="|Cmd|Result|\n|:---|:---|\n"
for _k,_v in _cmds.items():
    _md_table += (f"|`{_k}`|`{_v}`|\n")

# print(_md_table)

# display_markdown(_md_table)
display(Markdown(_md_table))


|Cmd|Result|
|:---|:---|
|`_task.get_reported_console_output()`|`['Process completed successfully']`|
|`_task._get_runtime_properties()`|`{'_CLEARML_TASK': True, '_exec_agent_hostname': 'bengio', 'CLEARML VERSION': 'clearml-1.16.4', 'CLI': 'test.py', 'progress': '0', 'platform': 'linux', 'python_version': '3.10.12', 'python_exec': '/home/blissfox-server/.clearml/venvs-builds/3.10/bin/python', 'OS': 'Linux-6.8.0-40-generic-x86_64-with-glibc2.35', 'processor': 'x86_64', 'cpu_cores': 56, 'memory_gb': 62.8, 'hostname': 'bengio', 'gpu_count': 1, 'gpu_type': 'NVIDIA TITAN X (Pascal)', 'gpu_memory': '12GB', 'gpu_driver_version': '535.183.01', 'gpu_driver_cuda_version': '12.2'}`|
|`_task.get_tags()`|`[]`|
|`_task.get_script()`|`{'working_dir': '.', 'entry_point': 'test.py', 'branch': 'main', 'repository': 'https://github.com/RR5555/clearml_demo.git'}`|
|`_task.comment`|`Auto-generated at 2024-09-17 12:23:52 UTC by blissfox-server@bengio`|
|`_task.status`|`completed`|
|`_task.get_parameters()`|`{}`|
|`_task.get_task_output_log_web_page(_task.id)`|`https://app.clearml.blissfox.xyz/projects/*/experiments/5812d0aa235f4858aa1e6c77925f0d3c/output/log`|
|`_task.name`|`clearml_test`|
|`_task.id`|`5812d0aa235f4858aa1e6c77925f0d3c`|
|`_task._project_name`|`('ce6d125a264149bc89e8e62fb1dd0d7a', None)`|
|`_task.get_output_destination()`|`None`|


Output example:

<div markdown=1 style=" overflow-x: auto;">

|Cmd|Result|
|:---|:---|
|`_task.get_reported_console_output()`|`['Process completed successfully']`|
|`_task._get_runtime_properties()`|`{'_CLEARML_TASK': True, '_exec_agent_hostname': 'bengio', 'CLEARML VERSION': 'clearml-1.16.4', 'CLI': 'test.py', 'progress': '0', 'platform': 'linux', 'python_version': '3.10.12', 'python_exec': '/home/blissfox-server/.clearml/venvs-builds/3.10/bin/python', 'OS': 'Linux-6.8.0-40-generic-x86_64-with-glibc2.35', 'processor': 'x86_64', 'cpu_cores': 56, 'memory_gb': 62.8, 'hostname': 'bengio', 'gpu_count': 1, 'gpu_type': 'NVIDIA TITAN X (Pascal)', 'gpu_memory': '12GB', 'gpu_driver_version': '535.183.01', 'gpu_driver_cuda_version': '12.2'}`|
|`_task.get_tags()`|`[]`|
|`_task.get_script()`|`{'working_dir': '.', 'entry_point': 'test.py', 'branch': 'main', 'repository': 'https://github.com/RR5555/clearml_demo.git'}`|
|`_task.comment`|`Auto-generated at 2024-09-17 12:23:52 UTC by blissfox-server@bengio`|
|`_task.status`|`completed`|
|`_task.get_parameters()`|`{}`|
|`_task.get_task_output_log_web_page(_task.id)`|`https://app.clearml.blissfox.xyz/projects/*/experiments/5812d0aa235f4858aa1e6c77925f0d3c/output/log`|
|`_task.name`|`clearml_test`|
|`_task.id`|`5812d0aa235f4858aa1e6c77925f0d3c`|
|`_task._project_name`|`('ce6d125a264149bc89e8e62fb1dd0d7a', None)`|
|`_task.get_output_destination()`|`None`|
</div>

In [11]:
print(_task.data)

<tasks.Task: {
    "id": "5812d0aa235f4858aa1e6c77925f0d3c",
    "name": "clearml_test",
    "user": "0f358a174884f2c34ef7def624d3e1a0",
    "company": "d1bd92a3b039400cbafc60a7a5b1e52b",
    "type": "training",
    "status": "completed",
    "comment": "Auto-generated at 2024-09-17 12:23:52 UTC by blissfox-server@bengio",
    "created": "2024-09-17 12:23:52.352000+00:00",
    "started": "2024-09-17 12:23:54.108000+00:00",
    "completed": "2024-09-17 12:28:29.926000+00:00",
    "active_duration": 275,
    "project": "ce6d125a264149bc89e8e62fb1dd0d7a",
    "output": {},
    "execution": {
        "queue": "6b55ab8e3be4482eb53acfed3e4037a5",
        "parameters": {},
        "model_desc": {},
        "model_labels": {},
        "artifacts": []
    },
    "container": {
        "image": "rr5555/dith:cuda12.0-agent",
        "arguments": "",
        "setup_shell_script": ""
    },
    "models": {
        "input": [],
        "output": []
    },
    "script": {
        "binary": "python",


In [ ]:
print(dir(_task.data))

In [65]:
# https://stackoverflow.com/a/50215584/8612123

def print_table(data, cols, wide):
    '''Prints formatted data on columns of given width.'''
    n, r = divmod(len(data), cols)
    pat = '{{:{}}}'.format(wide)
    line = '\n'.join(pat * cols for _ in range(n))
    last_line = pat * r
    print(line.format(*data))
    print(last_line.format(*data[n*cols:]))

In [66]:
print_table(dir(_task.data), 4, len(dir(_task.data))//4+1)

__class__                 __delattr__               __dict__                  __dir__                   
__doc__                   __eq__                    __format__                __ge__                    
__getattribute__          __gt__                    __hash__                  __init__                  
__init_subclass__         __le__                    __lt__                    __module__                
__ne__                    __new__                   __reduce__                __reduce_ex__             
__repr__                  __setattr__               __sizeof__                __str__                   
__subclasshook__          __weakref__               _data_props_list          _get_data_props           
_property_active_duration _property_comment         _property_company         _property_completed       
_property_configuration   _property_container       _property_created         _property_execution       
_property_hyperparams     _property_id              _pr

In [81]:
def get_instance_field(_obj, _class):
	return tuple(_field for _field in dir(_obj) if not callable(getattr(_obj, _field)) and isinstance(getattr(_obj, _field), _class))

In [85]:
_non_callable = tuple(_field for _field in dir(_task.data) if not callable(getattr(_task.data, _field)))
nb_fields = len(_non_callable)

print(f"Nb of non-callable fields: {nb_fields}/{len(_task.data.__dir__())}")

Nb of non-callable fields: 72/101


`Nb of non-callable fields: 72/101`

In [86]:
_types = tuple(type(getattr(_task.data, _field)) for _field in _non_callable)

In [89]:
_type_counts = dict((_type, _types.count(_type)) for _type in set(_types))

In [97]:
from pprint import pprint
pprint(_type_counts)
print(f"Fields (non-callable): {sum(_type_counts.values())}")

{<enum 'TaskStatusEnum'>: 2,
 <enum 'TaskTypeEnum'>: 2,
 <class 'str'>: 20,
 <class 'dict'>: 13,
 <class 'list'>: 4,
 <class 'int'>: 4,
 <class 'NoneType'>: 5,
 <class 'datetime.datetime'>: 14,
 <class 'clearml.backend_api.services.v2_23.tasks.TaskModels'>: 2,
 <class 'clearml.backend_api.services.v2_23.tasks.Execution'>: 2,
 <class 'clearml.backend_api.services.v2_23.tasks.Script'>: 2,
 <class 'clearml.backend_api.services.v2_23.tasks.Output'>: 2}
Fields (non-callable): 72


In [79]:
import datetime

In [82]:
_datetime_fields = get_instance_field(_task.data, datetime.datetime)
_str_fields = get_instance_field(_task.data, str)
_dict_fields = get_instance_field(_task.data, dict)
_list_fields = get_instance_field(_task.data, list)
_int_fields = get_instance_field(_task.data, int)
# pprint()
zip()

### Play with time

In [21]:
import datetime

In [68]:
datetime_fields = get_instance_field(_task.data, datetime.datetime)

print_table(datetime_fields, 1, len(datetime_fields)//1+1)

_property_completed
_property_created
_property_last_change
_property_last_update
_property_last_worker_report
_property_started
_property_status_changed
completed      
created        
last_change    
last_update    
last_worker_report
started        
status_changed 



In [22]:
print(f"`type(_task.data.created)`: {type(_task.data.created)}")
print(f"Created at: {_task.data.created}")

`type(_task.data.created)`: <class 'datetime.datetime'>
Created at: 2024-09-17 12:23:52.352000+00:00


To compare dates in `datetime.datetime`, we can use:
* Now: `datetime.datetime.now(tz=_task.data.created.tzinfo)`
* Deltas (7 days): `datetime.timedelta(days=7)`
* 7 days ago: `datetime.datetime.now(tz=_task.data.created.tzinfo) - datetime.timedelta(days=7)`

In [24]:
print(f"Now: {datetime.datetime.now(tz=_task.data.created.tzinfo)}")
print(f"7 days ago: {datetime.datetime.now(tz=_task.data.created.tzinfo) - datetime.timedelta(days=7)}")

Now: 2024-09-26 13:57:00.594638+00:00
7 days ago: 2024-09-19 13:57:00.594898+00:00


In [27]:
print(f"Was the last exp. created in the last 7 days: {'Yes' if (_task.data.created> datetime.datetime.now(tz=_task.data.created.tzinfo) - datetime.timedelta(days=7)) else 'No'}")

Was the last exp. created in the last 7 days: No


In [37]:
def get_exps_from_last(_list, _days):
	return tuple(_task for _task in _list if _task.data.created> datetime.datetime.now(tz=_task.data.created.tzinfo) - datetime.timedelta(days=_days))

In [38]:
print(f"Exp. in the last two months: {len(get_exps_from_last(completed_task_list, 60))}")
print(f"Exp. in the last year: {len(get_exps_from_last(completed_task_list, 365))}")

Exp. in the last two months: 1
Exp. in the last year: 9


# Launch an exp. from a private repo.

We will:
* Launch an agent/worker on a brand new queue: it will wait for a load (created a container? Or container only under load?)
* Launch a task from the CLI calling a script from a public GitHub repo
* Stop the agent

## Create a Personal Access Token (PAT)

<details>
<summary>Click to expand</summary>

To create a PAT, heads to `Settings`>`Developer settings`>`Personnal access tokens`>`Fine-grained tokens`, or click on the link below:\
https://github.com/settings/personal-access-tokens/new

This will lead you to this page:\
<img src="../img/new_PAT0.png" style="">\
Enter the `Token name`, `Expiration` & `Description`.

Select the repos you want the token to apply for:\
<img src="../img/new_PAT1.png" style="">

Choose the `Permissions`:
<img src="../img/new_PAT2.png" style="">

I only added the `Contents`:`Access: Read-only` (in addition to the default compulsory Metadata access):\
<img src="../img/new_PAT3.png" style="">

You can check out the permissions you will give in `Overview`:\
<img src="../img/new_PAT4.png" style="">\
Then confirm by `Generate token`.

You can now copy the token to where you will need it (you can only do it now, if you did not, you will have to regenerate the token, to copy the new one):\
<img src="../img/new_PAT5.png" style="">

</details>

## Launch the exp. from the private repo

To pass down the PAT either:
* configure in the default clearml-agent configs
* pass it down at agent creation
* modify the PAT in the tmp config of the clearml-agent daemon

### PAT at agent creation

To pass the PAT down at agent creation:

In [ ]:
%%bash
CLEARML_AGENT_GIT_USER="rr5555" CLEARML_AGENT_GIT_PASS=`cat ../pat.txt` CLEARML_WORKER_NAME=clearml-agent-bengio CLEARML_WORKER_ID=clearml-agent-bengio:gpu1 clearml-agent daemon --detached --queue gpu_dithering --create-queue --gpus 1
clearml-task --project Demo --name clearml_test --script test.py --repo https://github.com/RR5555/clearml_demo_private.git --branch main --queue gpu_dithering --docker rr5555/dith:cuda12.0-agent
clearml-agent daemon --stop clearml-agent-bengio:gpu1

ClearML launch - launch any codebase on remote machine running clearml-agent
Creating new task
Repository Detected
{
  "repository": "https://github.com/RR5555/clearml_demo_private.git",
  "branch": "main",
  "working_dir": ".",
  "entry_point": "test.py"
}
Base docker image: {'image': 'rr5555/dith:cuda12.0-agent', 'args': None, 'bash_script': None}
New task created id=d12f6da8220c46269c939c2040e71f97
Task id=d12f6da8220c46269c939c2040e71f97 sent for execution on queue gpu_dithering
Execution log at: https://app.clearml.blissfox.xyz/projects/ce6d125a264149bc89e8e62fb1dd0d7a/experiments/d12f6da8220c46269c939c2040e71f97/output/log


### PAT modification in the tmp config of the clearml-agent daemon

Create the clearml-agent daemon:

In [66]:
%%bash
CLEARML_WORKER_NAME=clearml-agent-bengio CLEARML_WORKER_ID=clearml-agent-bengio:gpu0 clearml-agent daemon --detached --queue gpu_dithering --create-queue --gpus 0

Current configuration (clearml_agent v1.9.1, location: /home/blissfox-server/clearml.conf):
----------------------
agent.worker_id = clearml-agent-bengio:gpu0
agent.worker_name = clearml-agent-bengio
agent.force_git_ssh_protocol = false
agent.python_binary = /usr/bin/python3.10
agent.package_manager.type = pip
agent.package_manager.pip_version.0 = <20.2 ; python_version < '3.10'
agent.package_manager.pip_version.1 = <22.3 ; python_version >\= '3.10' and python_version <\= '3.11'
agent.package_manager.pip_version.2 = >\=23,<24.3 ; python_version >\= '3.12'
agent.package_manager.system_site_packages = false
agent.package_manager.force_upgrade = false
agent.package_manager.conda_channels.0 = pytorch
agent.package_manager.conda_channels.1 = conda-forge
agent.package_manager.conda_channels.2 = nvidia
agent.package_manager.conda_channels.3 = defaults
agent.package_manager.priority_optional_packages.0 = pygobject
agent.package_manager.torch_nightly = false
agent.package_manager.poetry_files_f

Change the GitHub PAT in the tmp clearml-agent daemon cfg file:

In [71]:
import glob
import json

_cfg_file = glob.glob('/tmp/.clearml_agent.*.cfg')[0]

with open(_cfg_file,'r') as f:
	data = json.load(f)

with open('../pat.txt', 'r') as f:
	data['agent']['git_pass']=f.readline()

with open(_cfg_file,'w') as f:
	json.dump(data, f, indent=4)


In [72]:
%%bash
clearml-task --project Demo --name clearml_test --script test.py --repo https://github.com/RR5555/clearml_demo_private.git --branch main --queue gpu_dithering --docker rr5555/dith:cuda12.0-agent

ClearML launch - launch any codebase on remote machine running clearml-agent
Creating new task
Repository Detected
{
  "repository": "https://github.com/RR5555/clearml_demo_private.git",
  "branch": "main",
  "working_dir": ".",
  "entry_point": "test.py"
}
Base docker image: {'image': 'rr5555/dith:cuda12.0-agent', 'args': None, 'bash_script': None}
New task created id=bd37773fcb254a19b6915ab8279ec1a3
Task id=bd37773fcb254a19b6915ab8279ec1a3 sent for execution on queue gpu_dithering
Execution log at: https://app.clearml.blissfox.xyz/projects/ce6d125a264149bc89e8e62fb1dd0d7a/experiments/bd37773fcb254a19b6915ab8279ec1a3/output/log


In [73]:
%%bash
clearml-agent daemon --stop clearml-agent-bengio:gpu0

Terminating clearml-agent worker_id=clearml-agent-bengio:gpu0 pid=247263


## How to fail

Trying to directly run the experiment will **fail** as the repo is private and no PAT for it is passed:

In [ ]:
%%bash
CLEARML_WORKER_NAME=clearml-agent-bengio CLEARML_WORKER_ID=clearml-agent-bengio:gpu0 clearml-agent daemon --detached --queue gpu_dithering --create-queue --gpus 0
clearml-task --project Demo --name clearml_test --script test.py --repo https://github.com/RR5555/clearml_demo_private.git --branch main --queue gpu_dithering --docker rr5555/dith:cuda12.0-agent
clearml-agent daemon --stop clearml-agent-bengio:gpu0

It should also be noted that passing it down to `clearml-task` either as env. vars or docker env. vars will fail too:

In [1]:
# %%bash
# CLEARML_WORKER_NAME=clearml-agent-bengio CLEARML_WORKER_ID=clearml-agent-bengio:gpu0 clearml-agent daemon --detached --queue gpu_dithering --create-queue --gpus 0
# CLEARML_AGENT_GIT_USER="rr5555" CLEARML_AGENT_GIT_PASS=`cat ../pat.txt` clearml-task --project Demo --name clearml_test --script test.py --repo https://github.com/RR5555/clearml_demo_private.git --branch main --queue gpu_dithering --docker rr5555/dith:cuda12.0-agent
# clearml-agent daemon --stop clearml-agent-bengio:gpu0

In [ ]:
# %%bash
# CLEARML_WORKER_NAME=clearml-agent-bengio CLEARML_WORKER_ID=clearml-agent-bengio:gpu0 clearml-agent daemon --detached --queue gpu_dithering --create-queue --gpus 0
# clearml-task --project Demo --name clearml_test --script test.py --repo https://github.com/RR5555/clearml_demo_private.git --branch main --queue gpu_dithering --docker rr5555/dith:cuda12.0-agent --docker_args "-v CLEARML_AGENT_GIT_USER:rr5555 -v CLEARML_AGENT_GIT_PASS:`cat ../pat.txt`"
# clearml-agent daemon --stop clearml-agent-bengio:gpu0

> Note that this latest cell will expose in clear your PAT in the cell output and in the ClearML GUI.

# Pipelines

```bash
make launch-agents
make pipeline-from-functions
make pipeline-from-decorator

make create-tasks-for-pipeline
make pipeline-from-tasks

make stop-agents
```

https://clear.ml/docs/latest/docs/configs/clearml_conf/

agent.docker_use_activated_venv (bool)